src/state/slices/opportunitiesSlice/resolvers/cosmosSdk/index.ts

```diff
@@ -0,0 +1,60 @@
+import { cosmosChainId, fromAccountId, osmosisChainId } from '@shapeshiftoss/caip'
+import type { Account, CosmosSdkBaseAdapter, CosmosSdkChainId } from '@shapeshiftoss/chain-adapters'
+import { getChainAdapterManager } from 'context/PluginProvider/chainAdapterSingleton'
+import { logger } from 'lib/logger'
+import { isFulfilled, isRejected, isSome } from 'lib/utils'
+import type { ReduxState } from 'state/reducer'
+import { selectFeatureFlags, selectPortfolioAccounts } from 'state/slices/selectors'
+
+import type { GetOpportunityIdsOutput } from '../../types'
+import type { OpportunityIdsResolverInput } from '../types'
+import { makeUniqueValidatorAccountIds } from './utils'
+
+const moduleLogger = logger.child({ namespace: ['opportunities', 'resolvers', 'cosmosSdk'] })
+
+export const cosmosSdkOpportunityIdsResolver = async ({
+  reduxApi,
+}: OpportunityIdsResolverInput): Promise<{
+  data: GetOpportunityIdsOutput
+}> => {
+  const state = reduxApi.getState() as ReduxState
+
+  const { CosmosSdkOpportunitiesAbstraction } = selectFeatureFlags(state)
+  if (!CosmosSdkOpportunitiesAbstraction) {
+    return { data: [] }
+  }
+
+  const chainAdapters = getChainAdapterManager()
+  const portfolioAccountIds = Object.keys(selectPortfolioAccounts(state))
+  const cosmosSdkAccountIds = portfolioAccountIds.filter(accountId =>
+    [cosmosChainId || osmosisChainId].includes(fromAccountId(accountId).chainId),
+  )
+  const cosmosSdkAccounts = await Promise.allSettled(
+    cosmosSdkAccountIds.map(accountId => {
+      const { account: pubKey, chainId } = fromAccountId(accountId)
+      const adapter = chainAdapters.get(
+        chainId,
+      ) as unknown as CosmosSdkBaseAdapter<CosmosSdkChainId>
+      // TODO: skip ChainIds
+      return adapter.getAccount(pubKey)
+    }, {} as Record<string, Promise<Account<CosmosSdkChainId>>>),
+  ).then(settledAccountsPromises =>
+    settledAccountsPromises
+      .map(settledAccount => {
+        if (isRejected(settledAccount)) {
+          moduleLogger.error(settledAccount.reason, 'Error fetching Cosmos SDK account')
+          return undefined
+        }
+        if (isFulfilled(settledAccount)) return settledAccount.value
+
+        return undefined // This will never happen, a Promise is either rejected or fullfilled
+      })
```

Comment:
i feel like i saw some fancy await all settled util from apotheosis somewhere, but i can't recall where?

Reply:
Hmm can't grep fu anything in web but agree we def. should add it, this is very verbosey